### PR TITLE
Bluetooth: TMAP: Add role validation for bt_tmap_register

### LINKED ISSUE #60495: LE audio TMAP: Add role verification
**Is your enhancement proposal related to a problem? Please describe.**
want to discuss if we should add any abstraction layers for the different TMAP roles in tmap driver?

each TMAP role requires some special settings and low layer profile procedures.
image
but there is no more implementation for these in tmap.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/audio/tmap.c

we do these in demo apps:
https://github.com/zephyrproject-rtos/zephyr/pull/60288/commits/b926b294610a255eb81079a68764e3a486848e95#diff-cc104b7b2a9cb82bee5f3b6d56b8f3a54b4d2660731f7546c26bbd2eca310f5f
https://github.com/zephyrproject-rtos/zephyr/pull/56686/commits/f4f0f003a4e6cb3d88a7049a0c59625588eced32#diff-48dd3e98c744416052457825e6733baa10cfc8d0bfa0644aee65dddab49317c4

**Describe the solution you'd like**
add more apis to simplify the TMAP role usages.

### PR DESCRIPTION
**Pull request by Thalley**
Add validation that the device supports the role
set by bt_tmap_register.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/60495

### CODE REVIEW

File: subsys/bluetooth/audio/tmap.c [hunk 1]
```diff
@@ -155,10 +160,61 @@ static ssize_t read_role(struct bt_conn *conn,
...
+	}
+
+	if ((role & BT_TMAP_ROLE_CG) != 0 && !IS_ENABLED(CONFIG_BT_TMAP_CG_SUPPORTED)) {
+		LOG_DBG("Device does not support the CG role");
+
+		return false;
+	}
+
+	if ((role & BT_TMAP_ROLE_CT) != 0 && !IS_ENABLED(CONFIG_BT_TMAP_CT_SUPPORTED)) {
+		LOG_DBG("Device does not support the CY role");
```
Comment (kruithofa) [inline subsys/bluetooth/audio/tmap.c:176]:
```suggestion
		LOG_DBG("Device does not support the CT role");
```

File: subsys/bluetooth/audio/tmap.c [hunk 2]
```diff
@@ -155,10 +160,61 @@ static ssize_t read_role(struct bt_conn *conn,
...
+	}
+
+	return true;
+}
+
 int bt_tmap_register(enum bt_tmap_role role)
 {
 	int err;
 
+	CHECKIF(!valid_tmap_role(role)) {
```
Comment (kruithofa) [inline subsys/bluetooth/audio/tmap.c:212]: This is not specific for this PR, but also shows up other places where `CHECKIF` is used, and will be compiler dependent:
If the KCONFIG option `NO_RUNTIME_CHECKS` is set to 'y' a compiler may emit the following warning: `static bool valid_tmap_role defined but not used`

suggestions for fixing this?

**Review by kruithofa — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/audio/tmap.c (hunk 2 above)
Comment (Thalley) [reply subsys/bluetooth/audio/tmap.c:212]: Looking at the macro it is defined as

```c
#elif defined(CONFIG_NO_RUNTIME_CHECKS)
#define CHECKIF(...) \
	if (0)
#else
```

With `if (0)` the code is still compiled, but not linked, and should thus not have any issues. Will leave as is

Comment (kruithofa) [reply subsys/bluetooth/audio/tmap.c:212]: Yes, the code compiles and works fine as is, the only issue is that we get a compiler warning, that shouldn't be there.

**Review by kruithofa — DISMISSED** [review]

File: subsys/bluetooth/audio/Kconfig.tmap [hunk 3]
```diff
@@ -4,6 +4,23 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 #
+config BT_TMAP_CG_SUPPORTED
+	def_bool BT_CAP_INITIATOR && BT_CAP_COMMANDER && BT_BAP_UNICAST_CLIENT && BT_AUDIO_RX && BT_AUDIO_TX && BT_VCP_VOL_CTLR && BT_TBS
```
Comment (kartben) [inline subsys/bluetooth/audio/Kconfig.tmap:8]: would be nice to wrap this very long line?

File: subsys/bluetooth/audio/Kconfig.tmap [hunk 4]
```diff
@@ -4,6 +4,23 @@
...
+
+config BT_TMAP_CT_SUPPORTED
+	def_bool BT_CAP_ACCEPTOR && BT_BAP_UNICAST_SERVER
+
+config BT_TMAP_UMS_SUPPORTED
+	def_bool BT_CAP_INITIATOR && BT_CAP_COMMANDER && BT_BAP_UNICAST_CLIENT && BT_AUDIO_TX && BT_VCP_VOL_CTLR && BT_MCS
```
Comment (kartben) [inline subsys/bluetooth/audio/Kconfig.tmap:14]: ditto

**Review by larsgk — APPROVED** [review]

**Review by kartben — APPROVED** [review]

**Pull request merged**